src/Features/Core/Portable/UnusedReferences/UnusedReferencesRemover.cs

```diff
@@ -131,19 +143,42 @@ internal static class UnusedReferencesRemover
 
             foreach (var reference in references)
             {
-                // We will look at the compilation assemblies brought in directly by the
-                // references to see if they are used.
-                if (!reference.CompilationAssemblies.Any(usedAssemblyFilePaths.Contains))
+                if (reference.ReferenceType == ReferenceType.Project)
                 {
-                    // None of the assemblies brought into this compilation are in the
-                    // used assemblies list, so we will consider the reference unused.
-                    unusedReferencesBuilder.Add(reference);
-                    continue;
+                    // Since we only know project references by their CompilationReference which
+                    // does not include the full output path. We look only at the file name of the
+                    // compilation assembly and compare it with our list of used project assembly names.
+                    var projectAssemblyFileNames = reference.CompilationAssemblies
+                        .SelectAsArray(assemblyPath => Path.GetFileName(assemblyPath));
+
+                    // We will look at the project assemblies brought in directly by the
+                    // references to see if they are used.
+                    if (!projectAssemblyFileNames.Any(usedProjectFileNames.Contains))
+                    {
+                        // None of the project assemblies brought into this compilation are in the
+                        // used assemblies list, so we will consider the reference unused.
+                        unusedReferencesBuilder.Add(reference);
+                        continue;
+                    }
+
+                    // Remove the project file name now that we've identified it.
+                    usedProjectFileNames.ExceptWith(projectAssemblyFileNames);
```

Comment:
Does this need to happen when we add to the `unusedReferenceBuilder` as well?

Reply:
We don't because the check that adds a project to the `unusedReferenceBuilder` ensures that `usedProjectFileNames` doesn't contain any of the `projectAssemblyFileNames`.